Enable Clang static analyzer build.

coryan opened the pull request:
The changes are pretty simple, except for the tweaks to speed up
the build.  This is a fix for #10.

mbrukman commented on bigtable/CMakeLists.txt:
```diff
@@ -211,6 +211,14 @@ if (FORCE_SANITIZER_ERRORS)
 endif (FORCE_SANITIZER_ERRORS)
 
+option(FORCE_STATIC_ANALYZER_ERRORS
+    "If set enable tests that force errors detected by the static analyzer."
```
s/set enable/set, enable/

coryan replied:
Done.

mbrukman commented on ci/build-docker.sh:
```diff
@@ -22,10 +22,23 @@ set -eu
-cmake -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" ${CMAKE_FLAGS:-} ..
+CMAKE_COMMAND="cmake"
+if [ "${SCAN_BUILD}" = "yes" ]; then
+    CMAKE_COMMAND="scan-build cmake"
```
It's 2-space indent elsewhere in the file (below).

coryan replied:
Done.

mbrukman commented on ci/build-docker.sh:
```diff
@@ -22,10 +22,23 @@ set -eu
+# otherwise the static analyzer finds issues in them, and there is no way to
+# ignore them.
+if [ "${SCAN_BUILD}" = "yes" ]; then
+    make -j ${NCPU} -C bigtable googleapis
```
It's 2-space indent elsewhere in the file (below).

coryan replied:
Done.

mbrukman commented on ci/build-linux.sh:
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
      --build-arg CMAKE_FLAGS="${CMAKE_FLAGS:-}" \
      -f "ci/Dockerfile.${DISTRO}" .
+
+if [ "${SCAN_BUILD:-}" = "yes" ]; then
```
2-space indent here as well, please — it's now codified in our `.editorconfig` so it's official! ;-)

coryan replied:
Yeah, and my IDE seems to forget about it every other day... grrrr.

coryan commented:
PTAL.

We should wait until the PR finishes to see how long the builds take, we might want to disable the static analysis for PRs.

mbrukman commented:
SGTM on disabling static analysis for PRs, just run it on the `master` branch, no need to slow down PR review/merge process.

mbrukman commented on ci/build-linux.sh:
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
+if [ "${SCAN_BUILD:-}" = "yes" ]; then
+    sudo docker run --rm -it --volume $PWD:/d ${IMAGE}:tip \
+	 bash -c 'if [ ! -z "$(ls -1d /tmp/scan-build-* 2>/dev/null)" ]; then cp -r /tmp/scan-build-* /d/scan-build-output; fi'
+    if [ -r scan-build-output/index.html ]; then
```
This should be indented at the level of the previous line, no?

coryan replied:
`bash` is an argument to the `sudo docker` command, I thought we were indenting those by 2?
Anyway, changed the code to make this more readable.

mbrukman replied:
Sorry, misread, you're right.

mbrukman commented on ci/build-linux.sh:
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
+    sudo docker run --rm -it --volume $PWD:/d ${IMAGE}:tip \
+	 bash -c 'if [ ! -z "$(ls -1d /tmp/scan-build-* 2>/dev/null)" ]; then cp -r /tmp/scan-build-* /d/scan-build-output; fi'
+    if [ -r scan-build-output/index.html ]; then
+        echo -n $(tput setaf 1)
```
2-space indent?

coryan replied:
Fixed.

mbrukman commented on ci/build-linux.sh:
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
+	 bash -c 'if [ ! -z "$(ls -1d /tmp/scan-build-* 2>/dev/null)" ]; then cp -r /tmp/scan-build-* /d/scan-build-output; fi'
+    if [ -r scan-build-output/index.html ]; then
+        echo -n $(tput setaf 1)
+	echo "scan-build detected errors.  Please read the log for details."
```
I don't follow the indentation here. Should all the `echo` lines indent together?

coryan replied:
This was a mess, fixed.

mbrukman commented on ci/build-linux.sh:
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
+	echo "subdirectory."
+        exit 1
+    else
+        echo -n $(tput setaf 2)
```
2-space indent in the `else` clause.

coryan replied:
Fixed.

coryan commented:
PTAL.  I think I addressed all your comments, but I have rebased so many times I might have missed some.  Oh, and we need to wait for it to pass the tests, of course.

coryan commented:
I am going to close this PR because the build is still too slow, it just went from 37 minutes to 50+ minutes and aborted on my branch.

mbrukman commented:
Per offline conversation, please resurrect the part to allow folks to manually run this from the CLI, but leave out the part that automatically runs it on Travis CI — it will still be useful for us in the future to be able to easily run this on our workstations, even if it can't run automatically on Travis CI.
